src/twisted/python/_shellcomp.py

```diff
@@ -157,6 +158,8 @@ def __init__(self, options, cmdName, file):
         self.options = options
         self.cmdName = cmdName
         self.file = file
+        if file and ioType(file) == unicode:
```

Comment:
can the `file` ever be false ? 

Maybe we need a comment here to explain why we go with this strange file.buffer thing.

I would prefer to have the docstring updated to specify that `file` should be a BytesIO.

Then the caller of ZshBuilder should take care of callling it with the right IO.

So the fix should be in `shellComplete(config, cmdName, words, shellCompFile)` or in the caller of shellComplete

In this way, we will only fix the input stream in a single place and it will be fixed at the entry point, and not somewhere in the middle of this processing.

Reply:
There are some tests which do things like:
```
argGen = _shellcomp.ZshArgumentsGenerator(opts, 'ace', None)
```
so apparently that is allowed.

Reply:
After going through other parts of the Twisted code, especially Twisted written before Python 3, that when an argument of C{file} is specified, it really means a C{file} with bytes-semantics, such as BytesIO.  That's not clear until you dig into things.